tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,513 +15,568 @@
+ public:
   static constexpr int dataSize = 256;
 
-  testConnection(
-      [&](std::shared_ptr<transport::Connection> conn) {
-        std::shared_ptr<CpuContext> ctx = GetParam()->makeContext("server");
-        auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
-
-        // Initialize with sequential values.
-        std::vector<uint8_t> data(dataSize);
-        std::iota(data.begin(), data.end(), 0);
-
-        // Perform send and wait for completion.
-        std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
-        std::future<Error> sendFuture;
-        std::tie(descriptorFuture, sendFuture) =
-            sendWithFuture(channel, CpuTensor{data.data(), data.size()});
-        Error descriptorError;
-        TDescriptor descriptor;
-        std::tie(descriptorError, descriptor) = descriptorFuture.get();
-        EXPECT_FALSE(descriptorError) << descriptorError.what();
-        peers_->send(PeerGroup::kClient, descriptor);
-        Error sendError = sendFuture.get();
-        EXPECT_FALSE(sendError) << sendError.what();
-
-        peers_->done(PeerGroup::kServer);
-        peers_->join(PeerGroup::kServer);
-
-        ctx->join();
-      },
-      [&](std::shared_ptr<transport::Connection> conn) {
-        std::shared_ptr<CpuContext> ctx = GetParam()->makeContext("client");
-        auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
-
-        std::vector<uint8_t> data(dataSize);
-
-        // Perform recv and wait for completion.
-        auto descriptor = peers_->recv(PeerGroup::kClient);
-        std::future<Error> recvFuture = recvWithFuture(
-            channel, descriptor, CpuTensor{data.data(), dataSize});
-        Error recvError = recvFuture.get();
-        EXPECT_FALSE(recvError) << recvError.what();
-
-        // Validate contents of vector.
-        for (auto i = 0; i < data.size(); i++) {
-          EXPECT_EQ(data[i], i);
-        }
-
-        peers_->done(PeerGroup::kClient);
-        peers_->join(PeerGroup::kClient);
-
-        ctx->join();
-      });
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+    DataWrapper<TTensor> wrappedData(data);
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) =
+        sendWithFuture(channel, wrappedData.tensor());
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) = descriptorFuture.get();
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kClient, descriptor);
```

Comment:
This sounds like a C++ thing I'm unaware of, but why do we suddenly need `this->`?

Reply:
Damned C++, this is ADL all over again. I'm thankful I've learned about this the gentle way though...